### DISCUSSION TITLE
RequiresCompilation flag set to True for VisualBasicValue/Reference

### DISCUSSION THREAD

**Discussion by jsandip57** [UiPath/corewf#205] [Q&A]
To set the context, We are porting our application to .NetCore and currently validating CoreWF, to replace the legacy WF by Microsoft for runtime execution of WorkFlows. So, in our case we need to support the existing InstanceData saved using .Net 4.8 WorflowEngine. When we try to use the old data to resume the workflow in CoreWF, we faced an error(ValidationException) stating there is a mismatch in the Activity.MemberOf.Members list loaded from the existing InstanceData versus the one loaded workflow directly. After debugging we figured out the RequiresCompilation flag is set to true for VisualBasicValue & VisualBasicReference, this led to extra Expressions Activities added to the Members list, which is not the case in .Net4.8. This flag is set to true only for CSharpValue/References in.Net4.8 code base. By flipping the flag and recompiling the dll, we are able to load the existing data without any issue, though we are not sure of any side effects.

Since we are in the final stage of validation, we need to know the reason for this change in flag value, as we need to stick to the actual releases from Nuget instead of recompiling every time with this change.

Here are the screenshots of the Members list in .Net48 and .Net Core, with EnvironmentLocationReference activities are the newly added members for all VisualBasicValue/Reference activity.

![Net48Members](https://user-images.githubusercontent.com/3939915/163825294-1120f43e-e694-4e11-8018-8810186054ad.JPG)
![NetCoreMembers](https://user-images.githubusercontent.com/3939915/163825303-fe33edc5-7343-4c59-a16c-be6140cf96e5.JPG)

Also find the below error message that we get, while loading old data without flipping the RequiresCompilation Flag
![Error_CoreWF](https://user-images.githubusercontent.com/3939915/163836941-3ab6d1db-606a-4a24-879a-5d2a992f6427.JPG)

**Comment by dmetzgar** [comment]
The reason for the change comes from how VB and C# expressions are done in CoreWF vs .NET Framework. In .NET Framework, VB expressions are handled by a native VB compiler that ships with the .NET Framework. But for C# in .NET Framework, there is no equivalent C# compiler that can be embedded. Instead, the C# expressions are compiled as part of an MSBuild step - hence the existence of the flag RequiresCompilation.

In CoreWF, we use Roslyn for both VB and C#. We use the RequiresCompilation flag to trigger the ScriptingAOTCompiler. This creates a single assembly with all the VB/C# expressions in it. Turning off RequiresCompilation means you won't get those expressions. 

Could you create a small repro of the issue?

**Comment by jsandip57** [comment]
Thanks, @dmetzgar for the response. Here([Workflow.zip](https://github.com/UiPath/CoreWF/files/8571463/Workflow.zip)) is the simple console application built using .Net48 and .NetCore, to demonstrate the issue.

Also, we tried a couple of things, and here are our findings,
1) For loading the activity, we use ActivityXamlServices.Load method with CompileExpressions set to false by default. In this case, while doing cacheMetaData in VisualBasicValue, 
![image](https://user-images.githubusercontent.com/3939915/165495529-6a31254f-ac62-4a94-a442-f2be2f5a184e.png)
since the Environment.CompileExpressions is set to false, _expressionTree will be prepared. While executing the VisualBasic expression, in this context 
![image](https://user-images.githubusercontent.com/3939915/165495780-cfd73dcd-15cb-48c8-945d-b29175d41e99.png)
_compiledExpression will be used instead of the _invoker, and that's how it's working for us. But if the CompileExpressions is set to true, initially then we are dependent on _invoker and it crashes. 
2) Also we tried to set up another console application to load Workflows with CSharp expressions, in the case of CSharpValue we don't prepare the expression upfront, we rely only on the _invoker. Here in this case when we set ComplieExpression to false while loading, we were getting the below error 
"Expression Activity type 'CSharpValue`1' requires compilation in order to run.  Please ensure that the workflow has been compiled.". 

The ScriptingAotCompiler is used only when the CompileExression is set to True while loading the activity, 
![image](https://user-images.githubusercontent.com/3939915/165503767-e84dbca8-7114-42f1-9b68-e89ff2f7bbd2.png)
and also RequiresCompilation in CSharpValue is set to true, else while performing Compile as in the below image, 
![image](https://user-images.githubusercontent.com/3939915/165504564-dcf62fbc-94c0-43db-a419-eef54aed018f.png)
RequiresCompilation will be false and compilation will be skipped, resulting in a crash.

Our understanding is, that both AOT and JIT use the same Roslyn compiler for handling expressions. In the case of VisualBasic, while loading activity, if the AOT is not handling the compilation then on performing CacheMetaData, JIT will be used to compile the expression and proceed further.

**Reply by dmetzgar** [reply]
Thanks, I'll have a look at the sample. I'm curious why the CompiledExpressionInvoker crashes.

Regarding the RequiresCompilation for C#, you're right that it's inconsistent since we're using Roslyn. VB and C# should operate the same way. We should even be allowed to mix VB and C# expressions in the same workflow, something not allowed in Framework. I'll get back to you when I have an update.

**Reply by jsandip57** [reply]
Hello @dmetzgar, Any update on the above query?
Please let me know in case of any other inputs. 

Thanks!
